include/deal.II/lac/sparse_direct.h

```diff
@@ -452,99 +452,146 @@ class SparseDirectUMFPACK : public EnableObserverPointer
+                   const double lowrank_threshold = 1e-8)
+        : blr_ucfs(blr_ucfs)
+        , lowrank_threshold(lowrank_threshold)
+      {}
+      bool   blr_ucfs;
+      double lowrank_threshold;
+    };
+
+    AdditionalData(const bool          output_details    = false,
+                   const bool          error_statistics  = false,
+                   const bool          blr_factorization = false,
+                   const BlockLowRank &blr               = BlockLowRank())
+      : output_details(output_details)
+      , error_statistics(error_statistics)
+      , blr_factorization(blr_factorization)
+      , blr(blr)
+    {}
+
+    bool output_details;
+    bool error_statistics;
+
+    bool         blr_factorization;
+    BlockLowRank blr;
+  };
+  /**
+   * Constructor, takes <tt>AdditionalData</tt> to control MUMPS execution.
+   */
+  SparseDirectMUMPS(const AdditionalData &additional_data = AdditionalData());
 
   /**
-   *
+   * Destructor
    */
-  void
-  copy_rhs_to_mumps(const Vector<double> &rhs);
+  ~SparseDirectMUMPS();
 
   /**
-   * Flags storing whether the function <tt>initialize ()</tt> has already
-   * been called.
+   * Exception
    */
-  bool initialize_called;
+  DeclException0(ExcInitializeAlreadyCalled);
 
-public:
   /**
-   * Declare type for container size.
+   * This function computes the LU factorization
+   * of the system's matrix <tt>matrix</tt> with the options given in the
+   * constructor's <tt>AdditionalData</tt>.
    */
-  using size_type = types::global_dof_index;
+  template <class Matrix>
+  void
+  initialize(const Matrix &matrix);
 
   /**
-   * Constructor
+   * A function in which the inverse of the matrix is applied to the input
+   * vector <tt>src</tt> and the solution is written into the output vector
+   * <tt>dst</tt>.
    */
-  SparseDirectMUMPS();
+  void
+  vmult(Vector<double> &dst, const Vector<double> &src) const;
 
   /**
-   * Destructor
+   * A function in which the inverse of the transposed matrix is applied to the
+   * input vector <tt>src</tt> and the solution is written into the output
+   * vector <tt>dst</tt>.
    */
-  ~SparseDirectMUMPS();
+  void
+  Tvmult(Vector<double> &dst, const Vector<double> &src) const;
 
   /**
-   * Exception
+   * A function that returns the ICNTL integer array from MUMPS.
+   * The ICNTL array contains integer control parameters for the MUMPS solver.
+   * ICNTL array is originally a fortran array, it's 1-based.
+   * To select the correct index use a macro like this: #define ICNTL(I)
+   * icntl[(I)-1] In the MUMPS documentation there is the description of each
+   * parameter of the array.
    */
-  DeclException0(ExcInitializeAlreadyCalled);
+  int *
+  get_icntl();
+
+private:
+#ifdef DEAL_II_WITH_MUMPS
+  mutable DMUMPS_STRUC_C id;
+#endif // DEAL_II_WITH_MUMPS
+
+  double                     *a;
+  mutable std::vector<double> rhs;
+  int                        *irn;
+  int                        *jcn;
```

Comment:
We should eventually update this to use `std::unique_ptr` instead of raw pointers.